packages/query-engine/source/controllers/query-engine-controller.ts

```diff
@@ -1,47 +1,86 @@
 import type { MySQLConfiguration, Resource, Query } from "@hypertool/common";
 
-import mysql from "mysql2/promise";
 import { QueryTemplateModel, ResourceModel } from "@hypertool/common";
-import { NextFunction } from "express";
+import { Knex, knex } from "knex";
 
 import type { ExecuteParameters } from "../types";
 
-const execute = async (
+const executeMySQL = async (
     queryRequest: ExecuteParameters,
-    next: NextFunction,
+    query: Query,
+    resource: Resource,
 ): Promise<any> => {
-    const query: Query = await QueryTemplateModel.findOne({
-        name: queryRequest.name,
-    }).exec();
+    const mySQLConfig: MySQLConfiguration = resource.mysql;
+    const config: Knex.Config = {
+        client: "mysql",
+        connection: {
+            host: mySQLConfig.host,
+            port: mySQLConfig.port,
+            user: mySQLConfig.databaseUserName,
+            password: mySQLConfig.databasePassword,
+            database: mySQLConfig.databaseName,
+        },
+    };
+    const knexInstance = knex(config);
 
-    const resource: Resource = await ResourceModel.findOne({
-        _id: query.resource,
-    }).exec();
+    if (typeof query.content === "object") {
+        const queryContent: any = query.content;
 
-    const mySQLConfig: MySQLConfiguration = resource.mysql;
+        let queryResult;
+        if (queryContent.hasOwnProperty("select") && queryContent.select) {
```

Comment:
The query building needs to happen in a tree traversal fashion, recursively. For example, the first stage handles what kind of operation: select, insert, delete, and so on. The second stage takes care of from clause. The third stage takes care of where clause.

Why?

1. Duplicate code will be eliminated.
2. The objects can become more complicated. For example, in the current design, you cannot have optional where clauses.
3. The code will be concise, neat, and of better quality.

Reply:
Agreed.

Reply:
We can have a query building login that uses tree traversal fashion, but non-recursive for now. Is that okay?